### PR TITLE
Remove copies in image and video implementation

### PR DESCRIPTION
**Pull request by luisremis**
Optimized video and image code, avoiding copies of cvmats. This increase performance of video read and several image operations by up to 25%.

### CODE REVIEW

**Comment by vishakha041** [comment]
I don't see new tests for checking the copy parameters. Are our current tests sufficient to make sure we are not losing image content anywhere? Our typical checks involve metadata verification. I guess that concern applies to the old code too but just thought of it now.

**Comment by vishakha041** [comment]
What are the pros/cons of merging the ImageData/Image classes? That code was reviewed well back when and this change kind of breaks consistency across how video and descriptors were structured. Is that not true?

**Comment by luisremis** [comment]
> What are the pros/cons of merging the ImageData/Image classes? That code was reviewed well back when and this change kind of breaks consistency across how video and descriptors were structured. Is that not true?

These changes have been moved to another branch and will be part of a future pull request

**Comment by luisremis** [comment]
> I don't see new tests for checking the copy parameters. Are our current tests sufficient to make sure we are not losing image content anywhere? Our typical checks involve metadata verification. I guess that concern applies to the old code too but just thought of it now.

We have tests for both, the metadata an pixels. We have many tests that check every single pixel of the image and video and compare it with direct OpenCV results. The copies in the images are tests through the video implementation that is using those copie/move constructors. All tests are passing.

File: src/vcl/ImageData.cc [hunk 1]
```diff
@@ -318,26 +330,34 @@ ImageData::ImageData(void* buffer, cv::Size dimensions, int cv_type)
...
+        }
+    }
 
     if ( img._tdb != NULL )
         _tdb = new TDBImage(*img._tdb);
     else
         _tdb = NULL;
 
     int start;
+    // TODO IS THIS REALLY NEEDED????
```
Comment (vishakha041) [inline src/vcl/ImageData.cc]: Its better to not have comments like that. Either you can resolve it or file an issue and remove the comment

File: src/vcl/ImageData.cc [hunk 2]
```diff
@@ -752,16 +775,28 @@ void ImageData::delete_object()
...
     _channels = cv_img.channels();
 
     _height = cv_img.rows;
-    _width = cv_img.cols;
+    _width  = cv_img.cols;
 
     _cv_type = cv_img.type();
 
-    _cv_img = cv_img.clone();
+    _cv_img = cv_img; // shallow copy
```
Comment (vishakha041) [inline src/vcl/ImageData.cc]: Probably a silly question but is this still thread safe? That would be hard to catch with our tests.

File: src/vcl/VideoData.cc [hunk 3]
```diff
@@ -137,7 +137,7 @@ void VideoData::Write::operator()(VideoData *video)
     std::vector<VCL::Image>& frames = video->get_frames();
 
     for (auto& frame : frames) {
-        outputVideo << frame.get_cvmat();
+        outputVideo << frame.get_cvmat(false);
```
Comment (vishakha041) [inline src/vcl/VideoData.cc]: I can see why the shallow copy would be important here!

File: include/vcl/Image.h [hunk 4]
```diff
@@ -86,6 +86,8 @@ namespace VCL {
          */
         Image(const cv::Mat &cv_img);
 
+        Image(cv::Mat &cv_img);
```
Comment (luisremis) [inline include/vcl/Image.h]: see if we can use a parameter instead

File: src/vcl/ImageData.cc [hunk 5]
```diff
@@ -381,7 +401,8 @@ void ImageData::operator=(const ImageData &img)
         std::shared_ptr<Operation> front = img._operations.front();
         if (front->get_type() == OperationType::READ) {
             start = 1;
-            copy_cv(cv::imread(img._image_id, cv::IMREAD_ANYCOLOR));
+            cv::Mat img_read = cv::imread(img._image_id, cv::IMREAD_ANYCOLOR);
```
Comment (luisremis) [inline src/vcl/ImageData.cc]: check if this hurts or not.

File: src/vcl/ImageData.cc (hunk 2 above)
Comment (luisremis) [reply src/vcl/ImageData.cc]: none of this is thread safe. these changes do not affect that.

**Comment by luisremis** [comment]
All comments applied, all tests passing, noticeable improvement on video performance by avoing copies. Same should be good images that is now avoiding multiple copies. looks good to go!

**Pull request merged**